x/btcstaking/keeper/finality_providers.go

```diff
@@ -113,6 +129,111 @@ func (k Keeper) SlashFinalityProvider(ctx context.Context, fpBTCPK []byte) error
+	if btcTip == nil {
+		return fmt.Errorf("failed to get current BTC tip")
+	}
+	fp.SlashedBtcHeight = btcTip.Height
+	k.BscKeeper.SetConsumerFinalityProvider(ctx, fp)
+
+	// Process all delegations for this consumer finality provider and record slashed events
+	err = k.HandleFPBTCDelegations(ctx, fpBTCPK, func(btcDel *types.BTCDelegation) error {
+		stakingTxHash := btcDel.MustGetStakingTxHash().String()
+		eventSlashedBTCDelegation := types.NewEventPowerDistUpdateWithSlashedBTCDelegation(stakingTxHash)
+		k.addPowerDistUpdateEvent(ctx, btcTip.Height, eventSlashedBTCDelegation)
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to handle BTC delegations: %w", err)
+	}
+
+	return nil
+}
+
+// PropagateFPSlashingToConsumers propagates the slashing of a finality provider (FP) to all relevant consumer chains.
+// It processes all delegations associated with the given FP and creates slashing events for each affected consumer chain.
+//
+// The function performs the following steps:
+//  1. Retrieves all BTC delegations associated with the given finality provider.
+//  2. Collects slashed events for each consumer chain using collectSlashedConsumerEvents:
+//     a. For each delegation, creates a SlashedBTCDelegation event.
+//     b. Identifies the consumer chains associated with the FPs in the delegation.
+//     c. Ensures that each consumer chain receives only one event per delegation, even if multiple FPs in the delegation belong to the same consumer.
+//  3. Sends the collected events to their respective consumer chains.
+//
+// Parameters:
+// - ctx: The context for the operation.
+// - fpBTCPK: The Bitcoin public key of the finality provider being slashed.
+//
+// Returns:
+// - An error if any operation fails, nil otherwise.
+func (k Keeper) PropagateFPSlashingToConsumers(ctx context.Context, fpBTCPK *bbn.BIP340PubKey) error {
+	// Map to collect events for each consumer
+	consumerEvents := make(map[string][]*types.BTCStakingConsumerEvent)
+	// Create a map to store FP to consumer ID mappings
+	fpToConsumerMap := make(map[string]string)
+
+	// Process all delegations for this finality provider and collect slashing events
+	// for each consumer chain. Ensures that each consumer receives only one event per
+	// delegation, even if multiple finality providers in the delegation belong to the same consumer.
+	err := k.HandleFPBTCDelegations(ctx, fpBTCPK, func(delegation *types.BTCDelegation) error {
+		consumerEvent := types.CreateSlashedBTCDelegationEvent(delegation)
+
+		// Track consumers seen for this delegation
+		seenConsumers := make(map[string]struct{})
+
+		for _, delegationFPBTCPK := range delegation.FpBtcPkList {
+			fpBTCPKHex := delegationFPBTCPK.MarshalHex()
+			consumerID, exists := fpToConsumerMap[fpBTCPKHex]
+			if !exists {
+				// If not in map, check if it's a Babylon FP or get its consumer
+				if _, err := k.GetFinalityProvider(ctx, delegationFPBTCPK); err == nil {
+					continue // It's a Babylon FP, skip
```

Comment:
we already know if the finality provider `fpBTCPK` is from the Babylon when querying [HasFinalityProvider](https://github.com/babylonlabs-io/babylon/blob/873f1232365573a97032037af4ac99b5e3fcada8/x/btcstaking/keeper/btc_delegators.go#L79)

It would reduce one read by FP in the list, can be addressed in a future task

Reply:
Added a TODO in the code